File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
"""
Time = O(n)
Space = O(1)
"""
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # Create to accumulate multiplication arrays
        res = [1] * len(nums)
        # right = [1] * len(nums)
        
        # O(n)
        for i in range(1, len(res)):
            res[i] = nums[i-1] * res[i-1]
            
        # O(n)
        product = 1
        for i in range(len(res)-1, -1, -1):
            res[i] *= product
            product *= nums[i]
            
        return res
        
        
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> accProduct(nums.size(), 1);

        for (int i = 1; i < accProduct.size(); i++) {
            accProduct[i] = accProduct[i - 1] * nums[i - 1];
        }

        int curProduct = 1;
        for (int i = accProduct.size() - 1; i >= 0; i--) {
            accProduct[i] *= curProduct;
            curProduct *= nums[i];
        }

        return accProduct;
    }
};